ENH: pd.Series.shift and .diff to accept a collection of numbers

skwirskj opened the pull request:
- [x] closes #44424 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

skwirskj committed:
Laid out intial plans and ideas for how to handle series and DataFrame shift functions to allow iterables to be passed as the period

skwirskj committed:
ENH: Added functionality to DataFrame and Series shift functions to take in a list for periods and then return a concatenated DataFrame of each consecutive shift in the periods list

skwirskj committed:
Merge remote-tracking branch 'upstream/master' into 44424-pd-shift-collection

Merging the master branch with my changes for Enhancement at issue 44424

skwirskj committed:
TST: GH44424 Added tests for shift methods

Created two new tests to test shifting data in Series and DataFrames with an iterable as the period parameter.

skwirskj committed:
Merge remote-tracking branch 'upstream/master' into 44424-pd-shift-collection

Merging after updating tests and resetting the generic.py file

skwirskj committed:
ENH: GH#44424 Updated core/generic.py to latest master

Reverted some changes that were made when originally finding a solution and then reset it to the latest version on the upstream master branch.

skwirskj committed:
ENH: GH#44424 Updated whatsnew to explain changes

skwirskj committed:
ENH: GH#44424 Added Series.shift call to whatsnew

skwirskj committed:
Merge remote-tracking branch 'upstream/master' into 44424-pd-shift-collection

phofl commented on pandas/core/frame.py:
```diff
@@ -5444,6 +5444,34 @@ def shift(
+
+            from pandas.core.reshape.concat import concat
+
+            for i in periods:
```
Don‘t do this. Collect the frames and concat once at the end.

skwirskj replied:
fixed

phofl commented on pandas/core/frame.py:
```diff
@@ -5444,6 +5444,34 @@ def shift(
+            from pandas.core.reshape.concat import concat
+
+            for i in periods:
+                if not isinstance(i, int):
```
Check all at once

skwirskj replied:
Is this in reference to the isinstance call? If I pass the whole list for shifts into the function and am checking for ints, then it would return false.

phofl replied:
Yes, and the error. If a List of strings is passed you would raise for the first. User fixes first, you raise for the second, etc. the error should display all invalid values at once

GitChein replied:
So we can find all the indices that don't contain ints, and then raise the typerror.  i.e. check if each is an int in a separate loop above, and indicate all indices that aren't ints to the user

phofl replied:
Yep

jreback commented on pandas/core/frame.py:
```diff
@@ -5444,6 +5444,34 @@ def shift(
     ) -> DataFrame:
         axis = self._get_axis_number(axis)
 
+        # Handle the case of multiple shifts
```
prob reasonable to add a suffix arg (and if None use this as a default)

skwirskj replied:
Thank you for the suggestion, I will add that in!

GitChein replied:
We added in the agument

`suffix=None`

and in our shift:

`.add_suffix(f"_{i}" if suffix is None else suffix) `

jreback commented on pandas/core/frame.py:
```diff
@@ -5444,6 +5444,34 @@ def shift(
+                        f"Value {i} in periods is not an integer, expected an integer"
+                    )
+
+                new_df = concat(
```
this is terribly inefficient 

append to a list and concat at the end

skwirskj replied:
Got it. This would be a list of DataFrames, correct?

jreback replied:
yep

skwirskj replied:
Thank you!

GitChein replied:
We now append to a list in the for loop

`new_df_list.append(
                        super()
                        .shift(periods=i, freq=freq, axis=axis, fill_value=fill_value)
                        .add_suffix(f"_{i}" if suffix is None else suffix) 
                    )`

and we concat at the end

`new_df = concat(new_df_list, axis=1)`

phofl commented on pandas/core/generic.py:
```diff
@@ -12,6 +12,7 @@
 from typing import (
     TYPE_CHECKING,
     Any,
+    AnyStr,
```
I assume something went wrong here?

GitChein replied:
We didn't really touch generic.py, we are confused as to what is going wrong here.  First time working with this library, so any context would be appreciated.

skwirskj replied:
To add to this, we followed the contributing guidelines instructions about merging upstream with our feature branch and it appears to have introduced the AnyStr addition to generic.py

phofl replied:
Looks like your latest merge fixed it

skwirskj committed:
Merge remote-tracking branch 'upstream/master' into 44424-pd-shift-collection

skwirskj committed:
ENH: GH#44424 DataFrame.shift performance updates

Updated how we are creating the new DataFrame after shifting to improve performance. Also reverted core/generic.py to master

skwirskj committed:
BUG: GH#44424 Formatting whatsnew

Attempting to fix whatsnew rst file

skwirskj committed:
Merge remote-tracking branch 'upstream/master' into 44424-pd-shift-collection

skwirskj commented:
@phofl @jreback I noticed that I am failing a check in CI/Checks. It is due to Series.shift expecting a series as output, but the issue requested that in the case of an utterable being passed for the period parameter, that Series.shift returns a DataFrame. Is there some way to get around this?

rhshadrach commented on pandas/core/series.py:
```diff
@@ -4939,6 +4939,16 @@ def _replace_single(self, to_replace, method: str, inplace: bool, limit):
     # error: Cannot determine type of 'shift'
     @doc(NDFrame.shift, klass=_shared_doc_kwargs["klass"])  # type: ignore[has-type]
     def shift(self, periods=1, freq=None, axis=0, fill_value=None) -> Series:
```
The return type should be `Series | DataFrame`. Making this change might cause other typing issues; e.g. if the code expects getting back a Series when calling `.shift(1)`, mypy will no longer be able to deduce this function is returning a Series even though it is.

If this occurs, you can type-hint `periods` as `int | Collection[int]` (collection is a sized iterable) and then use overloads. This way mypy will be able to deduce when it gets back a Series vs a DataFrame based on the input.

skwirskj replied:
Thank you for the help with this! I will change the expected to Series | DataFrame for shift and also type-hint periods.

If you have the time to answer this, does the "CI / checks" check use the type-hints to verify what type is being returned vs what should be returned?

rhshadrach replied:
It uses mypy. You should be able to run this yourself by going to the package directory and running `mypy pandas` from the command line.

skwirskj committed:
ENH: GH#44424 Added typing hints since shift can return series or dataframe now

jreback commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -186,6 +186,38 @@ representation of :class:`DataFrame` objects (:issue:`4889`).
+
+Usage within the :class:`DataFrame` class:
+
+.. ipython:: python
```
this is fine, but the main source of information is the doc-strings which need updating with examples & prose.

jreback commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -186,6 +186,38 @@ representation of :class:`DataFrame` objects (:issue:`4889`).
+
+The :meth:`DataFrame.shift` and :meth:`Series.shift` functions can take in an iterable, such as a list, for the period parameter. When an iterable is passed
+to either function it returns a :class:`DataFrame` object with all of the shifted rows or columns concatenated with one another.
+The function applies a shift designated by each element in the iterable. The resulting :class:`DataFrame` object's columns will retain the
```
you don't need this sentence to the end. you can say that the ``suffix`` parameter controls the names but don't have to fully explain here. that's the point of the doc-string.

jreback commented on pandas/core/frame.py:
```diff
@@ -5326,9 +5326,38 @@ def shift(
     ) -> DataFrame:
         axis = self._get_axis_number(axis)
 
+        # GH#44424 Handle the case of multiple shifts
```
this comment is not needed

jreback replied:
if axis==1 this should fail

jreback commented on pandas/core/frame.py:
```diff
@@ -5326,9 +5326,38 @@ def shift(
+            for i in periods:
+                if not isinstance(i, int):
+                    raise TypeError(
+                        f"Value {i} in periods is not an integer, expected an integer"
```
does this have tests?

jreback commented on pandas/core/frame.py:
```diff
@@ -5326,9 +5326,38 @@ def shift(
+        # GH#44424 Handle the case of multiple shifts
+        if is_list_like(periods):
+
+            new_df = DataFrame()
```
this is superfluous

jreback commented on pandas/core/frame.py:
```diff
@@ -5326,9 +5326,38 @@ def shift(
+
+            from pandas.core.reshape.concat import concat
+
+            new_df_list = []
```
use result and not new_df

jreback commented on pandas/core/series.py:
```diff
@@ -4938,7 +4939,18 @@ def _replace_single(self, to_replace, method: str, inplace: bool, limit):
+    ) -> Series | DataFrame:
+        # Handle the case of multiple shifts
+        if is_list_like(periods):
+            if len(periods) == 0:
```
is this tested? this should be an error i think

jreback commented on pandas/core/series.py:
```diff
@@ -4938,7 +4939,18 @@ def _replace_single(self, to_replace, method: str, inplace: bool, limit):
+            if len(periods) == 0:
+                return self
+
+            df = self.to_frame()
```
why are you doing this? shouldn't you just be concatting the result series?

jreback commented on pandas/tests/frame/methods/test_shift.py:
```diff
@@ -429,3 +429,23 @@ def test_shift_axis1_categorical_columns(self):
         tm.assert_frame_equal(result, expected)
+
+    def test_shift_with_iterable(self):
+        # GH#44424
```
what about error conditions

jreback commented on pandas/tests/series/methods/test_shift.py:
```diff
@@ -376,3 +377,15 @@ def test_shift_non_writable_array(self, input_data, output_data):
         tm.assert_series_equal(result, expected)
+
+    def test_shift_with_iterable(self):
+        # GH#44424
```
same question

jreback commented:
closing as stale if you want continue, pls ping

jona-sassenhagen commented:
@jreback I am considering taking over this PR, would that be ok with you?
